Synthesise lambda classes

smowton opened the pull request:
This scans invokedynamic instructions during class-loading, creating a synthetic type for each one that we can understand, and registers its constructor (which captures) and method (which passes captured and actual parameters to the real lambda function) as synthetic methods. These are then created on demand using the same mechanism as e.g. synthetic static initializer methods.

Needs tests, but the code is in reviewable state.

smowton committed:
Factor out create_stub_method_symbol

smowton committed:
Record class' instance methods against their type

This makes it much easier to find the type's methods without having to search the
whole symbol table.

smowton committed:
Implement lambda code synthesis

This scans invokedynamic instructions during class-loading, creating a synthetic type
for each one that we can understand, and registers its constructor (which captures)
and method (which passes captured and actual parameters to the real lambda function)
as synthetic methods. These are then created on demand using the same mechanism as
e.g. synthetic static initializer methods.

Prior to this change invokedynamic instructions were always stubbed to return null.
With this change they actually work, so you can verify programs that use lambdas,
such as assert ((MyInterface)(x -> x + 1))(1) == 2

tautschnig commented:
@smowton For Java-clueless people like myself: can you explain _why_ this change is desirable to have?

smowton commented:
At the moment `invokedynamic` instructions are always stubbed to return null. With this change they actually work, so you can verify programs that use lambdas, such as `assert (MyInterface)(x -> x + 1)(1) == 2`

tautschnig commented:
@smowton Thanks! Would you mind adding that info to the commit message?

thk123 commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -94,6 +94,35 @@ static void assign_parameter_names(
   }
 }
 
+void create_method_stub_symbol(
```
:pick: I think would be more natural if this returned the symbol, and elt the caller add to the symbol table

LAJW replied:
@thk123 This directly manipulates the symbol table, so returning/constness is not very relevant. I don't see a way to make this substantially better 😞

thk123 commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -2185,23 +2214,15 @@ void java_bytecode_convert_methodt::convert_invoke(
-    symbol_table.add(symbol);
+    create_method_stub_symbol(
+      invoked_method_id,
+      arg0.get(ID_C_base_name),
```
:bulb: Lots of the parameters are derived from `ID_C_base_name`, perhaps just pass that in and let the method do this substring nonsense

thk123 replied:
Or possibly even better - pass in arg0

thk123 commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -324,14 +324,15 @@ optionalt<symbolt> java_bytecode_convert_methodt::get_lambda_method_symbol(
 /// \param message_handler: A message handler to collect warnings.
 void java_bytecode_convert_method_lazy(
-  const symbolt &class_symbol,
+  symbolt &class_symbol,
```
:bulb: This should be unit tested

smowton commented:
@tautschnig done

smowton committed:
Move lambda test script to test_goto_functions.desc

We're about to add a real test of the same lambdas, so first we move what will be an
ancillary test out of the way

smowton committed:
Remove no-longer-applicable explanation of symex-driven-loading-xfail

Test no longer has this tag, but the excuse for having it was erroneously retained.

smowton committed:
Lambda tests: use local interface

The models library doesn't have java.util.function.* yet, so for the time being we must use
our own versions of these interfaces.

smowton committed:
Expand existing lambda test to check that lambdas behave as expected

smowton commented:
Added an expansion of the `jbmc/lambda1` test to actually check behaviour

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: a94e7a9909ea856ff8d6cda127253685d9587c75).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/115779638

smowton committed:
Ignore lambdas whose target interfaces have multiple methods

This currently accounts for those with default methods, and those which re-declare as
abstract methods inherited from Object, such as equals(Object). In the future we should
distinguish and disregard these methods as candidates for the intended lambda method.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 87a435fc9fee0561c6b59e4e1e00ae5a49f4e144).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/115786915

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 447de944b2eddaa85dc97b910118a18d49872b01).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/115796756

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 4b8837eb8435c5bd3b7f2f6fc5baa08d99c4ab75).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/115930962

smowton commented:
Pinging @thk123 @peterschrammel for review

peterschrammel commented on jbmc/src/java_bytecode/java_bytecode_convert_method.cpp:
```diff
@@ -2957,40 +2937,71 @@ code_blockt java_bytecode_convert_methodt::convert_astore(
-  pop(parameters.size());
+  if(!symbol_table.has_symbol(synthetic_class_name))
+  {
+    // We failed to parse the invokedynamic handle as a Java 8+ lambda;
```
Should we output a warning in this case?

smowton replied:
A debug message is already produced in `java_bytecode_parser.cpp` in this case. That's probably enough for now, as otherwise we would print a warning for every non-Java use of invokedynamic, which would be pretty user-unfriendly

peterschrammel commented on jbmc/src/java_bytecode/lambda_synthesis.cpp:
```diff
@@ -134,6 +134,18 @@ void create_invokedynamic_synthetic_classes(
                     << messaget::eom;
         continue;
       }
+      else if(implemented_interface_type.methods().size() != 1)
```
(commit message) distiniguish -> distinguish

peterschrammel reviewed (approved):
LGTM

thk123 commented on jbmc/src/java_bytecode/java_bytecode_language.cpp:
```diff
@@ -719,6 +720,33 @@ bool java_bytecode_languaget::typecheck(
+  // Now add synthetic classes for every invokedynamic instruction found (it
+  // makes this easier that all interface types and their methods have been
+  // created above):
+  {
```
:bulb: Maybe make this a function rather than unlabelled scope?

thk123 commented on jbmc/src/java_bytecode/java_bytecode_language.cpp:
```diff
@@ -719,6 +720,33 @@ bool java_bytecode_languaget::typecheck(
+  {
+    std::vector<irep_idt> function_symbols;
+
+    for(const auto &id_and_symbol : symbol_table)
```
:bulb: This looks like a natural fit for a filter

thk123 commented on jbmc/src/java_bytecode/java_bytecode_language.cpp:
```diff
@@ -719,6 +720,33 @@ bool java_bytecode_languaget::typecheck(
+    {
+      if(id_and_symbol.second.type.id() != ID_code)
+        continue;
+      auto cmb = method_bytecode.get(id_and_symbol.first);
```
:no_entry_sign: I don't know what `cmb` stands - please de-acro-fy
:question: I don't know what `id_and_symbol.first` is likely to be, and I don't think direct irep access using a dynamic string is a good idea - can we do something a bit better here

thk123 commented on jbmc/src/java_bytecode/java_bytecode_language.cpp:
```diff
@@ -719,6 +720,33 @@ bool java_bytecode_languaget::typecheck(
+      function_symbols.push_back(id_and_symbol.first);
+    }
+
+    for(const auto &id : function_symbols)
```
:bulb: `id` isn't an informative name, perhaps `method_id`

thk123 commented on jbmc/src/java_bytecode/java_bytecode_language.cpp:
```diff
@@ -719,6 +720,33 @@ bool java_bytecode_languaget::typecheck(
+
+    for(const auto &id_and_symbol : symbol_table)
+    {
+      if(id_and_symbol.second.type.id() != ID_code)
```
:pick: `if(can_cast_type<code_typet>(id_and_symbol.second.type))`

thk123 commented on jbmc/src/java_bytecode/java_bytecode_language.cpp:
```diff
@@ -1031,6 +1059,15 @@ static void notify_static_method_calls(
           expr_dynamic_cast<symbol_exprt>(fn_call->function());
         needed_lazy_methods->add_needed_method(fn_sym.get_identifier());
       }
+      else if(
```
:question: What is this block about and how come it is appearing as part of this commit?

smowton replied:
Separate commit sounds good, they're added because we now make side-effect-function-call expressions to avoid a pointless temporary (the expression is equivalent to a `code_function_callt` with a symbol on the LHS followed by immediately returning that symbol)

thk123 commented on jbmc/src/java_bytecode/java_bytecode_language.cpp:
```diff
@@ -1031,6 +1059,15 @@ static void notify_static_method_calls(
         needed_lazy_methods->add_needed_method(fn_sym.get_identifier());
       }
+      else if(
+        it->id() == ID_side_effect &&
```
:pick: `expr_try_dynamic_cast`

thk123 commented on jbmc/src/java_bytecode/java_bytecode_language.cpp:
```diff
@@ -1136,6 +1173,14 @@ bool java_bytecode_languaget::convert_single_method(
           get_pointer_type_selector(),
           get_message_handler());
       break;
+    case synthetic_method_typet::INVOKEDYNAMIC_CAPTURE_CONSTRUCTOR:
```
:question: What are these? can they do in a seperate commit with justification?

thk123 replied:
These are added because these are new kinds of methods to convert.

thk123 commented on jbmc/src/java_bytecode/lambda_synthesis.cpp:
```diff
@@ -0,0 +1,440 @@
+  {
+    if(!strcmp(bytecode_info[instruction.bytecode].mnemonic, "invokedynamic"))
+    {
+      // invokedynamic will be called with operands that should be stored in a
```
Seems like this big picture documentation should live in the doxygen for the function?

thk123 commented on jbmc/src/java_bytecode/lambda_synthesis.cpp:
```diff
@@ -0,0 +1,440 @@
+      // invoke instruction will return null when eventually converted by
+      // java_bytecode_convert_method.
+
+      const auto &method_symbol = ns.lookup(method_identifier);
```
You might find the immediately invoked lambda a useful pattern here. Or pull out a little static function to get the `class_type` from the `method_identifier` since all other variables are used only here

thk123 replied:
Actually even `class_type` isn't used outside of the next block

thk123 commented on jbmc/src/java_bytecode/lambda_synthesis.cpp:
```diff
@@ -0,0 +1,440 @@
+
+      if(implemented_interface_type.get_is_stub())
+      {
+        log.debug() << "ignoring invokedynamic at " << method_identifier
```
:question: Seems like this should be a `log.warning()`?

smowton replied:
Disagree, this would be a very noisy high-priority warning, suggest this should only be upgraded as/when stub classes stop being the norm

thk123 commented on jbmc/src/java_bytecode/lambda_synthesis.cpp:
```diff
@@ -0,0 +1,440 @@
+      irep_idt synthetic_class_name =
+        lambda_synthetic_class_name(method_identifier, instruction.address);
+
+      java_class_typet synthetic_class_type;
```
:bulb: Seems like this could be pulled out into a separate function too `create_synthetic_lambda_class` or something?

thk123 commented on jbmc/src/java_bytecode/lambda_synthesis.cpp:
```diff
@@ -0,0 +1,440 @@
+
+      // Add the class fields:
+
+      java_class_typet::componentt base_field;
```
:question: There must be code that already creates this component somewhere else - can it be reused?

thk123 commented on jbmc/src/java_bytecode/lambda_synthesis.cpp:
```diff
@@ -0,0 +1,440 @@
+        synthetic_class_type.components().emplace_back(std::move(new_field));
+      }
+
+      // Create constructor symbol:
```
:bulb: It seems like you've already named this function!

thk123 commented on jbmc/src/java_bytecode/lambda_synthesis.cpp:
```diff
@@ -0,0 +1,440 @@
+      set_declaring_class(constructor_symbol, synthetic_class_name);
+      symbol_table.add(constructor_symbol);
+
+      // Create implemented method symbol:
```
:bulb: And this one!

thk123 commented on jbmc/src/java_bytecode/lambda_synthesis.cpp:
```diff
@@ -0,0 +1,440 @@
+  }
+}
+
+static const symbolt &get_or_create_method_symbol(
```
:question: How come this function is needed here specifically? Seems like we already have method that do this?

thk123 replied:
:question: I still don't understand what this method is for or why is lambda specific

thk123 commented on jbmc/src/java_bytecode/lambda_synthesis.cpp:
```diff
@@ -0,0 +1,440 @@
+  return symbol_table.lookup_ref(identifier);
+}
+
+codet get_invokedynamic_synthetic_constructor(
```
:no_entry_sign: Personally find `get` a little misleading in the name here, would `synthesize` be more accurate? Perhaps adding a doxygen summary would help

smowton replied:
`create_` then?

thk123 commented on jbmc/src/java_bytecode/lambda_synthesis.cpp:
```diff
@@ -0,0 +1,440 @@
+    parameters.at(0).get_identifier(), parameters.at(0).type());
+  const dereference_exprt deref_this(this_param);
+
+  // Call super-constructor (always java.lang.Object):
```
:bulb: seems like another function

thk123 commented on jbmc/src/java_bytecode/lambda_synthesis.cpp:
```diff
@@ -0,0 +1,440 @@
+
+    if(parameter.get_this())
+      continue;
+
```
:bulb: Probably worth throwing some invariants here that the members and the parameters all exist, their types are correct etc?

thk123 commented on jbmc/src/java_bytecode/lambda_synthesis.cpp:
```diff
@@ -0,0 +1,440 @@
+    ++field_iterator;
+  }
+
+  return std::move(result);
```
:question: Why move? In this case is a local variable, won't RVO kick in anyway?

smowton replied:
Needed when you return `code_blockt` from a function whose type says it produces `codet`, as there will be an implicit copy- or move-constructor call to perform the conversion. Some Clang versions actually warn (or with `-Werror` give a fatal error) when you don't do this (or if the types are the same, if you *do*)

thk123 commented on jbmc/src/java_bytecode/lambda_synthesis.cpp:
```diff
@@ -0,0 +1,440 @@
+  return std::move(result);
+}
+
+codet get_invokedynamic_synthetic_method(
```
Same comment - I think `get` is not a good term for this

thk123 commented on jbmc/src/java_bytecode/lambda_synthesis.cpp:
```diff
@@ -0,0 +1,440 @@
+    param_symbol.base_name = parameter.get_base_name();
+    param_symbol.mode = ID_java;
+    param_symbol.type = parameter.type();
+    symbol_table.add(param_symbol);
```
:question: Why is this happening so late, is that normal?

smowton replied:
Matches what we do with normal (non-synthetic) methods

thk123 commented on jbmc/src/java_bytecode/lambda_synthesis.cpp:
```diff
@@ -0,0 +1,440 @@
+
+  if(return_type != empty_typet())
+  {
+    result.add(code_returnt(side_effect_expr_function_callt(
```
:question: why is this a `side_effect_expr_function_callt` and not just a regular `code_function_callt`? Why do both methods even exist?

thk123 replied:
A: to avoid declaring, storing and then returning a value... meh

smowton replied:
Yeah, one's an expression, the other a statement

thk123 commented on jbmc/regression/jbmc/lambda1/BiFunction.java:
```diff
@@ -0,0 +1,6 @@
+
+interface BiFunction<From1, From2, To> {
```
:question: Why? I guess these interfaces don't exist in core models, but might be easier to just add them?

smowton replied:
That's up to them, I just wanted a free-standing test

thk123 commented on jbmc/src/java_bytecode/lambda_synthesis.cpp:
```diff
@@ -134,6 +134,18 @@ void create_invokedynamic_synthetic_classes(
       }
+      else if(implemented_interface_type.methods().size() != 1)
+      {
+        log.debug() << "ignoring invokedynamic at " << method_identifier
```
:pick: Seems like should definitely be a warning?

smowton replied:
Default methods, Object methods, and stub types all produce false positives at the moment, so suggest this is still debug level until those issues are addressed.

thk123 commented on jbmc/regression/jbmc/lambda1/Lambdatest.java:
```diff
@@ -76,17 +74,54 @@ public boolean custom(Integer i) {
+  public static void main(String[] args, int unknown) {
+      Lambdatest lt = new Lambdatest();
+      if(unknown == 0)
+        assert lt.g(1.0f, 2, (x, y) -> x.intValue() + y) == 3; // should succeed
```
:pick: I know you didn't name these functions, but if we could maybe rename some of these so less impossible to know what is being tested here `g` -> `executeLambda`

thk123 commented on jbmc/regression/jbmc/lambda1/Lambdatest.java:
```diff
@@ -76,17 +74,54 @@ public boolean custom(Integer i) {
+      else if(unknown == 1)
+        assert lt.g(1.0f, 2, (x, y) -> x.intValue() + y) == 4; // should fail
+      else if(unknown == 2)
+        assert lt.f(1.0f, 2, 3) == 9; // should succeed
```
:pick: `f` -> `capture_local`

thk123 commented on jbmc/regression/jbmc/lambda1/Lambdatest.java:
```diff
@@ -76,17 +74,54 @@ public boolean custom(Integer i) {
+      else if(unknown == 3)
+        assert lt.f(1.0f, 2, 3) == 10; // should fail
+      else if(unknown == 4)
+        assert lt.i(2) == 10; // should succeed
```
`i` -> `capture_local_primitive`

thk123 commented on jbmc/regression/jbmc/lambda1/Lambdatest.java:
```diff
@@ -76,17 +74,54 @@ public boolean custom(Integer i) {
+      else if(unknown == 5)
+        assert lt.i(2) == 11; // should fail
+      else if(unknown == 6)
+        assert lt.j(3) == 30; // should succeed
```
`j` -> `capture_member`

thk123 commented on jbmc/regression/jbmc/lambda1/Lambdatest.java:
```diff
@@ -76,17 +74,54 @@ public boolean custom(Integer i) {
+      else if(unknown == 7)
+        assert lt.j(3) == 31; // should fail
+      else if(unknown == 8)
+        assert lt.k(4) == 40; // should succeed
```
`k` -> `capture_reference_member`

thk123 commented on jbmc/regression/jbmc/lambda1/Lambdatest.java:
```diff
@@ -76,17 +74,54 @@ public boolean custom(Integer i) {
+        assert lt.k(4) == 41; // should fail
+      else if(unknown == 10)
+      {
+        assert lt.l(5) == 70; // should succeed
```
`l` -> `side_effect_lambda`

thk123 reviewed (dismissed):
:no_entry_sign: Missing tests for:
 - capturing a primitive array :heavy_check_mark: 
 - capturing a reference array :heavy_check_mark: 
 - capturing an object factory produced array/reference type (reference done, but can't see array?)
 - returning something captured (sort of tested - but might be good to verify the reference equality holds?)
 - assigning a lambda from an interface to an interface:
```java
Func<A, B> x = (a) -> return a;
Func<A, B> y = x;
y.Apply(1);
```
 - lambda that capture members of outer classes :heavy_check_mark: 
 - lamdbas that are initalized in `<clinit>` (missing but in fairness GH had rendered this bullet point incomprehensible)

The tests are also quite difficult to figure out what is tested due to the cryptic names (not yours but still)

There are also unit tests that should, with a bit of tweaking, now work inside `jbmc/unit/java_bytecode/java_bytecode_convert_method` - could these be turned on?

My biggest complaint on the code is there are huge chunks of straight line code that would be easier to digest if they were split up into smaller functions. You're half way there - you've broken them up with whitespace and stuck a comment that is essentially a function name above them.

smowton committed:
Add tests for as-yet unsupported lambda types

Commit:
Add more lambda tests

Commit:
Group create_invokedynamic_synthetic_classes into scopes

Commit:
Expose only name/id instead of the class

Commit:
Convert named scopes into functions

Commit:
Convert lambdas into functions

Commit:
Pull out function create_implemented_method_symbol for invokedynamic conversion

Commit:
Combine create_synthetic_symbol and create_synthetic_type

Commit:
Decrease nesting and spacing name functions consistently

Commit:
Move the comment outside of the function

Commit:
Remove controversial get_ prefix

Because nobody could decie between: get_, create_, synthesize_, etc...

Commit:
Replace includes with forward declarations

Commit:
Get rid of the block scope and cmb variable

Commit:
Use can_cast instead of id comparison

smowton commented:
@lajw reviewing my own PR, realised the lambda1 test is needlessly changed from tab size 4 to 2. I'd fix it myself but I'm on the road until Sunday with only a phone.

thk123 commented on jbmc/src/java_bytecode/lambda_synthesis.cpp:
```diff
@@ -66,8 +66,8 @@ void create_invokedynamic_synthetic_classes(
 {
-  messaget log(message_handler);
-  namespacet ns(symbol_table);
+  messaget log{message_handler};
```
:pick: Putting this sort of changes in the same commit increases the burden for the reviewer, if possible put them in a separate commit in future

thk123 commented on jbmc/src/java_bytecode/lambda_synthesis.cpp:
```diff
@@ -120,10 +123,11 @@ void create_invokedynamic_synthetic_classes(
-        ns.lookup(implemented_interface_tag.get_identifier());
-      const java_class_typet &implemented_interface_type =
-        to_java_class_type(implemented_interface_symbol.type);
+      const java_class_typet &implemented_interface_type = [&] {
```
:question: I'll trust your answer - but is storing the result of function in a reference (albeit const ref) a good idea?

LAJW replied:
`const ref` extends the lifespan of the value on the right. The `&` is superfluous. This is the result of me trying to not replace all types with `const auto` 😁

thk123 commented on jbmc/src/java_bytecode/lambda_synthesis.cpp:
```diff
@@ -118,6 +118,96 @@ static optionalt<irep_idt> interface_method_id(
+  constructor_symbol.type = constructor_type;
+  set_declaring_class(constructor_symbol, synthetic_class_name);
+  return constructor_symbol;
+      }
```
:pick: This commit `Convert lambdas into functions` seems to need clang formatting?

thk123 commented on jbmc/src/java_bytecode/lambda_synthesis.cpp:
```diff
@@ -206,7 +206,56 @@ static symbolt create_constructor_symbol(
-      }
+}
+
+symbolt create_implemented_method_symbol(
```
:no_entry_sign: Please make the commit message more informative:
`Conver the last function ` -> `Pull out function create_implemented_method_symbol for invokedynamic conversion`

thk123 commented on jbmc/src/java_bytecode/lambda_synthesis.cpp:
```diff
@@ -118,7 +118,7 @@ static optionalt<irep_idt> interface_method_id(
 }
 
-const java_class_typet synthetic_class_type(
+symbolt synthetic_class_symbol(
```
:green_heart:

thk123 commented on jbmc/src/java_bytecode/java_bytecode_language.cpp:
```diff
@@ -723,20 +723,11 @@ bool java_bytecode_languaget::typecheck(
-    }
-
-    for(const auto &id : function_symbols)
+    const auto &id = id_and_symbol.first;
```
:pick: `id` is better than `cmb`, but might be easier to just use `symbol.name` whether `id` is used, since then obvious is just the id of the symbol we're talking abotu

thk123 commented on jbmc/src/java_bytecode/lambda_synthesis.cpp:
```diff
@@ -0,0 +1,499 @@
+                  << " with unknown handle type" << messaget::eom;
+      continue;
+    }
+    const auto &implemented_interface_tag = to_struct_tag_type(
```
:bulb: Name suggestion: `functional_interface_tag`

thk123 reviewed (commented):
Thanks for all the clean up Lukasz - this is looking much better. There are a couple of tests still missing:

- capturing an object factory produced array type (reference done, but not array?):

```java
int testFunction(Foo[] inputArr) {
  lambda = () -> return inputArr.length;
  lambda.execute();
}
```

- returning something captured (sort of tested - but might be good to verify the reference equality holds?):

```java
int testFunction() {
  Object reference = new Object();
  lambda = () -> return reference;
  Object o2 = lambda.execute();
  assert o2 == reference;
}
```

- assigning a lambda from an interface to an interface:
```java
Func<A, B> x = (a) -> return a;
Func<A, B> y = x;
y.Apply(1);
```

- lamdbas that are initalized in <clinit> (missing but in fairness GH had rendered this bullet point incomprehensible)

```java
static Func<A, A> x = (a) -> return a;
int testFunction() {
  A a = new A();
  A result = x.Apply(a);
  assert(a == result);
}
```

It would also be really good to tidy up the variable names in the existing test as I commented in the original review, but this can be done in a follow up PR. 

Finally could we have some rudimentary tests in TG with a bump pointing at this branch.

Commit:
Slice type_symbolt on creation

Change required by clang builds

smowton committed:
Avoid misleading special characters within lambda synthetic class names

Enough code in cbmc and its users relies on method names having the form
'java::package.Class.method:(type)' that the existing method for synthesising
lambda class names, which could yield things like java::synthetic_lambda$main:()V$0.<init>,
could cause problems when that colon character in the middle of the name was mistaken
for the function name's trailing type information. Similarly we might end up
needing to append fake type information to the <init> and apply symbol names to
appear "normal" enough.

LAJW commented:
Test-gen update with extra tests: https://github.com/diffblue/test-gen/pull/3697
Edit: Rebased

Commit:
Rename implemneted_interface_tag to functional_interface_tag

Commit:
Use `symbol.name` as the id

Commit:
Rename method names in the lambda test

Commit:
Add reference equality and static initialized lambda check

Commit:
Add nondet capture test

LAJW commented:
Rebased
Reformatted
Added tests:
 - nondet capture
 - statically initialized lambda
 - test for checking reference equality of captured object

Test not added:
 - Assignment of lambda from one variable to the next, because it's implicitly there.

Commit:
Add nondet array primitive and reference tests

LAJW commented:
Nondet array tests (reference and primitive) added.

jeannielynnmoulton reviewed (approved):
Verifying that the tests that @thk123 requested have been added since he is on holiday.

LAJW commented:
Rebasing on top of latest develop

Commit:
Formatting

tautschnig commented:
@smowton @LAJW Two of the Travis/cmake configurations are still failing.

smowton commented:
@LAJW @tautschnig just rebased on current develop which should work around the dodgy goto-harness test.

tautschnig commented:
@smowton CI still seems unhappy, though now in different ways.

smowton committed:
Fix invokedynamic typing

smowton committed:
Sync test_goto_functions and test.desc classpaths

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: a316fd5be7b0baaa63da8857823041e313554984).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/122445309
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.
